[sdk] Initial implementation of a reflection-based PropertyValue deserializer

Zaid-Ajaj opened the pull request:
This PR implements a deserializer from `PropertyValue` into objects via reflection, to be used in native-provider implementations that operate on `PropertyValue` instead of the lower-level `Struct` and `Value` from gRPC. 

The implementation handles most commonly used primitive types and their containers: `string`, `int`, `bool`, `double`, enums, nullable types, arrays, immutable arrays, lists, dictionaries, immutable dictionaries, inputs, input list, input map and nested objects. 

When deserializing an object, if a non-nullable _primitive_ property is missing, an exception is thrown. However, this doesn't include arrays, dictionaries or lists: an empty container is constructed when the data is empty or missing. Also for strings, an empty string is created when deserializing from `Null` or the property is missing. This is to accommodate for the fact that gRPC treats null values as empty so I made the deserializer less strict about missing values.

Frassle commented:
> When deserializing an object, if a non-nullable primitive property is missing, an exception is thrown. However, this doesn't include arrays, dictionaries or lists: an empty container is constructed when the data is empty or missing. Also for strings, an empty string is created when deserializing from Null or the property is missing. This is to accommodate for the fact that gRPC treats null values as empty so I made the deserializer less strict about missing values.


Ooooh. I'm conflicted on this.
_I_ like strict systems, so I'd probably prefer the system to error on _any_ missing required property. But then you've got the fun of working out if something is "required" or not. Easy for value types (is it Nullable), but we'd probably need an attribute for reference types.

But most of pulumi works on a very loose, protobuf style interface where a missing field is the same as the zero value for that field. So to work well in practice with that ecosystem we might need this system to treat missing as zero value as well.

Zaid-Ajaj commented:
@Frassle I've refactored the serializer code into a `PropertyValueSerializer` class which exposes two _instance_ methods as we discussed:
 - `Task<PropertyValue> Serialize(object? value)`
 - `T Deserialize<T>(PropertyValue value)`

Removed all encoder/decoder APIs and the associated handling of attributes. 

The only "extension point" right now is that users can use `PropertyValue` directly when serializing or deserializing when they want to handle the raw data directly (with the idea is the inferred schema later would be `any` when using `PropertyValue` directly) 

Would appreciate a second look 🙏

Zaid-Ajaj commented on sdk/Pulumi/Provider/PropertyValueSerializer.cs:
```diff
@@ -0,0 +1,747 @@
+                   + $"Expected {expected} instead at path {propertyPath}.";
+        }
+
+        public T Deserialize<T>(PropertyValue value)
```
> I've got a few cases I've been thinking of which I'm not sure if they should be 'errors' or bubbled up as extra values here. Mostly around unknowns and secrets.

You mean returning something like `Result<T, DeserializeError>`?

> But like if you have a property value Secret(Number(4)) and you try and pass that to Deserialize<double> what should we do?

Right now this will fail because `Deserialize<double>` is expecting `Number` but instead it gets `Secret`. A `Secret` number means that the value must be an output/input to allow _secretness_ . Using `Deserialize<Input<double>>` will work in this case. 

This makes me think that, actually `Secret(Number(4))` should not be possible to represent in the first place in the domain model. Instead maybe something like `Output(value=Number(4), secret=True)`?

Frassle replied:
> You mean returning something like Result<T, DeserializeError>?

Probably more like { Value: T?, Secret: bool }, T? because it might have been a nested unknown value. But at this point it's like this is basically just `Output<T>` so maybe Deserialise just has to deserialise into an `Output<T>`?

> This makes me think that, actually Secret(Number(4)) should not be possible to represent in the first place in the domain model. Instead maybe something like Output(value=Number(4), secret=True)?

Well both are valid at the moment, and providers/sdks do use both forms. Also for a provider potentially _any_ input could have been made secret (except for plains?) but then you start looking at something like Ian's Output<T> in provider SDKs to support that.

Also even if we decide that `Secret(Number(4))` is an error to deserialise into just `double` what if the secretness is nested? Like `Array([Secret(Number(1)), Number(2)])` deser'd into `Output<double[]>`. Even though the item is only expected to be a `double` not an output it feels safe to lift the secretness up to the nearest output value?

This is definitely something to think about, and probably get comments from the provider team about how they think this should look based on using the Go provider SDK.

Zaid-Ajaj replied:
>Also for a provider potentially any input could have been made secret (except for plains?)

The way I am thinking about this is that when you `Deserialize<double>` you are explicitly saying: "I am expecting a plain value from the raw wire data format". If the data contains secrets then using `double` wasn't the right choice. Users should then use `Input<double>`


> Deserialise just has to deserialise into an Output<T>?

Users would almost always want `T` from here, not the `Output<T>` when deserializing objects or values. In fact, they could just use `Deserialize<Input<T>>` and it will work for plain values as well, I believe

Zaid-Ajaj commented on sdk/Pulumi/Core/InputList.cs:
```diff
@@ -98,6 +98,15 @@ internal InputList<T> Clone()
+        /// Specialized overload used to construct an InputList from Input of an ImmutableArray.
+        /// This function is called using Reflection which is why it is marked internal.
+        /// </summary>
+        internal static InputList<T> FromInputList(Input<ImmutableArray<T>> values)
```
I guess we could, will do that instead of static methods 👍

Zaid-Ajaj replied:
Removed this and all added helper methods using from reflection APIs ✅ instead using existing implicit operators

Zaid-Ajaj commented on sdk/Pulumi/Provider/PropertyValue.cs:
```diff
@@ -401,7 +406,7 @@ public bool TryGetArray(out ImmutableArray<PropertyValue> value)
         }
 
-        public bool TryGetObject(out ImmutableDictionary<string, PropertyValue>? value)
+        public bool TryGetObject([System.Diagnostics.CodeAnalysis.NotNullWhen(true)] out ImmutableDictionary<string, PropertyValue>? value)
```
> Are these the only out parameters we need to care about?

Yeah I believe so, the others return struct so they can't be null

Frassle replied:
This is the sort of tiny change that if you put it into another PR I can whack approve and we can get it in right away :)

Zaid-Ajaj commented on sdk/Pulumi/Serialization/OutputData.cs:
```diff
@@ -37,5 +38,8 @@ public void Deconstruct(out X value, out bool isKnown, out bool isSecret)
             isSecret = IsSecret;
         }
+
+        public static OutputData<X> Create(ImmutableHashSet<Resource> resources, X value, bool isKnown, bool isSecret)
```
Removed this and all added helper methods using from reflection APIs ✅

Zaid-Ajaj commented on sdk/Pulumi/Provider/PropertyValue.cs:
```diff
@@ -1,9 +1,14 @@
 using System.Linq;
 using System.Collections.Generic;
 using System.Collections.Immutable;
+using System.Reflection;
```
Removed ✅

Zaid-Ajaj commented:
@Frassle refactored the API according to our discussion:
 - The `PropertyValueSerializer` is now marked with `[Obsolete("... use at your own risk ...")]` to start with
 - Deserialize now returns `Task<T>`
 - No longer returning default values for containers and is strict about what can be `null` and what cannot
 - Removed all helper methods that were being used from reflection, now relying on existing constructors and implicit operators

Frassle commented on sdk/Pulumi/Provider/PropertyValueSerializer.cs:
```diff
@@ -0,0 +1,751 @@
+
+            if (targetType.IsEnum)
+            {
+                var enumValue = Enum.GetName(targetType, value);
```
We write these as their names rather than values? That feels a bit surprising, again I'd copy the behavior from System.Text.Json for this and AFAIK by default it writes values.

Zaid-Ajaj replied:
Changed implementation so that enums write out their values instead names ✅ will need to remember that when doing schema-gen

Frassle commented on sdk/Pulumi/Provider/PropertyValueSerializer.cs:
```diff
@@ -0,0 +1,751 @@
+            var propertyType = property.PropertyType;
+            var propertyName = property.Name;
+            var inputAttribute =
+                propertyType
```
Shouldn't this attribute lookup be on the property itself, not the type.

Zaid-Ajaj replied:
Fixed and added tests ✅

Frassle commented on sdk/Pulumi/Provider/PropertyValueSerializer.cs:
```diff
@@ -0,0 +1,751 @@
+            }
+
+            var outputAttribute =
+                propertyType
```
Likewise

Frassle commented on sdk/Pulumi.Tests/Provider/PropertyValueTests.cs:
```diff
@@ -0,0 +1,310 @@
+        Assert.True(data.IsSecret);
+        Assert.True(data.IsKnown);
+    }
+}
```
We're missing tests of the Input/Output attributes to rename fields.

Zaid-Ajaj replied:
Added those ✅

Zaid-Ajaj commented on sdk/Pulumi/Provider/PropertyValueSerializer.cs:
```diff
@@ -0,0 +1,751 @@
+                return await Serialize(valueField);
+            }
+
+            if (targetType.IsArray)
```
Done ✅

Zaid-Ajaj commented on sdk/Pulumi/Provider/PropertyValueSerializer.cs:
```diff
@@ -0,0 +1,751 @@
+    {
+        private object? DeserializeObject(ImmutableDictionary<string, PropertyValue> inputs, Type targetType, string[] path)
+        {
+            var instance = Activator.CreateInstance(targetType);
```
I've implemented constructor-based deserialization ✅ (see tests)

Zaid-Ajaj commented on sdk/Pulumi.Tests/Provider/PropertyValueTests.cs:
```diff
@@ -0,0 +1,310 @@
+    [Fact]
+    public async Task DeserializingBasicArgsWorks()
+    {
+        var basicArgs = await Deserialize<BasicArgs>(new Dictionary<string, object?>
```
Rewrote the tests such that we now build the property map by hand before using the deserializer ✅

Zaid-Ajaj commented on sdk/Pulumi.Tests/Provider/PropertyValueTests.cs:
```diff
@@ -0,0 +1,310 @@
+
+public class PropertyValueTests
+{
+    async Task<T> Deserialize<T>(Dictionary<string, object?> inputs)
```
> Feels like tests should be explicitly building PropertyValues for this sort of thing

Refactored the tests to do this ✅

Zaid-Ajaj commented on sdk/Pulumi.Tests/Provider/PropertyValueTests.cs:
```diff
@@ -0,0 +1,310 @@
+{
+    async Task<T> Deserialize<T>(Dictionary<string, object?> inputs)
+    {
+        var serializer = new PropertyValueSerializer();
```
Done ✅

Zaid-Ajaj commented on sdk/Pulumi/Provider/PropertyValueSerializer.cs:
```diff
@@ -0,0 +1,751 @@
+            return propertyName;
+        }
+
+        public async Task<PropertyValue> Serialize(object? value)
```
Done ✅

Zaid-Ajaj commented on sdk/Pulumi/Provider/PropertyValueSerializer.cs:
```diff
@@ -0,0 +1,751 @@
+            {
+                case string stringValue:
+                    return new PropertyValue(stringValue);
+                case int integer:
```
Added short, ushort, long, ulong, float, decimal, and byte ✅ we can add anything I've missed later on 👍

Frassle reviewed (approved):
This is looking about right to me, lets merge it so we can start experimenting with uses.
